File: tools/runtime/patch_runtime.py
#!/usr/bin/env python3
"""
Generalized runtime patching tool for replacing Go runtime functions.

This script patches ELF binaries to redirect function calls from one function
to another. It's used to replace Go runtime functions with bare-metal implementations.

Usage:
    patch_runtime.py <elf_file> <old_func1> <new_func1> [<old_func2> <new_func2> ...]

Example:
    patch_runtime.py kernel.elf runtime.gcWriteBarrier2 runtime.gcWriteBarrier2
"""

import struct
import sys
import subprocess
import os
import re

def find_bin_dir():
    """Find the bin directory containing target-* tools."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    # tools/runtime -> tools -> project root -> bin
    project_root = os.path.dirname(os.path.dirname(script_dir))
    bin_dir = os.path.join(project_root, 'bin')
    return bin_dir

def setup_env():
    """Set up environment with bin directory in PATH."""
    bin_dir = find_bin_dir()
    env = os.environ.copy()
    if 'PATH' in env:
        env['PATH'] = bin_dir + ':' + env['PATH']
    else:
        env['PATH'] = bin_dir
    return env

def get_text_section_info(elf_path, env):
    """Get .text section file offset and virtual address."""
    result = subprocess.run(['target-readelf', '-S', elf_path],
                          capture_output=True, text=True, env=env, check=True)
    
    for line in result.stdout.split('\n'):
        if '.text' in line and 'PROGBITS' in line:
            parts = line.split()
            try:
                for i, part in enumerate(parts):
                    if part == 'PROGBITS' and i + 2 < len(parts):
                        vaddr_str = parts[i+1].strip()
                        offset_str = parts[i+2].strip()
                        vaddr = int(vaddr_str, 16)
                        offset = int(offset_str, 16)
                        return (offset, vaddr)
            except (ValueError, IndexError) as e:
                print(f"Warning: Error parsing line: {line}, error: {e}")
                continue
    
    raise RuntimeError("Could not find .text section")

def find_symbol_address(elf_path, symbol_name, env, prefer_type=None):
    """Find the address of a symbol using target-nm.
    
    Args:
        elf_path: Path to ELF file
        symbol_name: Name of symbol to find
        env: Environment with PATH set
        prefer_type: Preferred symbol type ('T' for global, 't' for local/weak)
                    If None, returns first match. If specified, tries preferred type first.
    """
    result = subprocess.run(['target-nm', elf_path],
                          capture_output=True, text=True, env=env, check=True)
    
    # nm output format: <address> <type> <name>
    # Collect all matches first
    matches = []
    for line in result.stdout.split('\n'):
        parts = line.split()
        if len(parts) >= 3:
            addr_str, sym_type, name = parts[0], parts[1], ' '.join(parts[2:])
            if name == symbol_name and sym_type in 'Tt':
                try:
                    addr = int(addr_str, 16)
                    matches.append((addr, sym_type))
                except ValueError:
                    continue
    
    if not matches:
        raise RuntimeError(f"Could not find symbol: {symbol_name}")
    
    # If preference specified, try to find matching type first
    if prefer_type:
        for addr, sym_type in matches:
            if sym_type == prefer_type:
                return addr
    
    # Return first match (or only match if no preference)
    return matches[0][0]

def find_call_sites(elf_path, target_addr, env):
    """Find all call sites (bl instructions) that call the target address."""
    result = subprocess.run(['target-objdump', '-d', elf_path],
                          capture_output=True, text=True, env=env, check=True)
    
    call_sites = []
    # Pattern: <address>: <instruction> <target_addr> <symbol_name>
    # Example: "27c2a4:	97ffca93 	bl	26ecf0 <runtime.gcWriteBarrier2>"
    pattern = re.compile(r'^\s*([0-9a-f]+):\s+([0-9a-f]{8})\s+bl\s+([0-9a-f]+)')
    
    for line in result.stdout.split('\n'):
        match = pattern.match(line)
        if match:
            call_addr_str, insn_str, target_str = match.groups()
            call_addr = int(call_addr_str, 16)
            target = int(target_str, 16)
            if target == target_addr:
                call_sites.append((call_addr, int(insn_str, 16)))
    
    return call_sites

def encode_bl_instruction(pc, target):
    """Encode a bl (branch with link) instruction.
    
    bl encoding: 0x94000000 | ((target - pc) >> 2) & 0x3ffffff
    """
    offset = target - pc
    if offset % 4 != 0:
        raise ValueError(f"Target address must be 4-byte aligned (offset={offset})")
    if not (-0x8000000 <= offset <= 0x7ffffff):
        raise ValueError(f"Branch offset out of range: {offset} (max ±128MB)")
    
    imm26 = (offset >> 2) & 0x3ffffff
    return 0x94000000 | imm26

def patch_call_site(data, file_offset_base, vaddr_base, call_vaddr, old_target, new_target):
    """Patch a single call site to redirect from old_target to new_target."""
    # Calculate file offset for the call instruction
    file_offset = file_offset_base + (call_vaddr - vaddr_base)
    
    if file_offset < 0 or file_offset >= len(data):
        print(f"Warning: Invalid file offset {file_offset} for call at 0x{call_vaddr:x}")
        return False
    
    # Read current instruction
    current_insn = struct.unpack_from('<I', data, file_offset)[0]
    
    # Verify it's a bl instruction
    if (current_insn & 0xfc000000) != 0x94000000:
        print(f"Warning: Instruction at 0x{call_vaddr:x} is not a bl instruction: 0x{current_insn:x}")
        return False
    
    # Decode current target to verify
    imm26 = current_insn & 0x3ffffff
    # Sign extend 26-bit immediate
    if imm26 & 0x2000000:
        imm26 |= 0xfc000000  # Sign extend
    current_offset = imm26 << 2
    current_target = call_vaddr + current_offset
    
    if current_target != old_target:
        print(f"Warning: Call at 0x{call_vaddr:x} targets 0x{current_target:x}, expected 0x{old_target:x}")
        # Continue anyway - maybe we still want to patch it
    
    # Encode new instruction
    try:
        new_insn = encode_bl_instruction(call_vaddr, new_target)
    except ValueError as e:
        print(f"Error: Cannot encode bl instruction for call at 0x{call_vaddr:x}: {e}")
        return False
    
    # Write the new instruction
    struct.pack_into('<I', data, file_offset, new_insn)
    
    print(f"Patched call at 0x{call_vaddr:x} (file offset 0x{file_offset:x}): "
          f"0x{old_target:x} -> 0x{new_target:x}")
    return True

def patch_runtime(elf_path, replacements):
    """
    Patch ELF binary to redirect function calls.
    
    Args:
        elf_path: Path to ELF file to patch
        replacements: List of (old_func_name, new_func_name) tuples
    """
    env = setup_env()
    
    # Get .text section info
    file_offset_base, vaddr_base = get_text_section_info(elf_path, env)
    print(f"Found .text section: vaddr=0x{vaddr_base:x}, file_offset=0x{file_offset_base:x}")
    
    # Read ELF file
    with open(elf_path, 'rb') as f:
        data = bytearray(f.read())
    
    total_patches = 0
    
    # Process each replacement
    for old_func, new_func in replacements:
        print(f"\nProcessing replacement: {old_func} -> {new_func}")
        
        try:
            # Find addresses
            # Old function: prefer weak/local symbols (type 't') from Go runtime
            # New function: prefer strong/global symbols (type 'T') from our code
            old_addr = find_symbol_address(elf_path, old_func, env, prefer_type='t')
            new_addr = find_symbol_address(elf_path, new_func, env, prefer_type='T')
            print(f"  Old function address: 0x{old_addr:x}")
            print(f"  New function address: 0x{new_addr:x}")
            
            # Find all call sites
            call_sites = find_call_sites(elf_path, old_addr, env)
            print(f"  Found {len(call_sites)} call site(s)")
            
            if not call_sites:
                print(f"  Warning: No call sites found for {old_func}")
                continue
            
            # Patch each call site
            for call_vaddr, _ in call_sites:
                if patch_call_site(data, file_offset_base, vaddr_base, call_vaddr, old_addr, new_addr):
                    total_patches += 1
        
        except RuntimeError as e:
            print(f"  Error: {e}")
            continue
    
    # Write patched file
    if total_patches > 0:
        with open(elf_path, 'wb') as f:
            f.write(data)
        print(f"\nSuccessfully patched {total_patches} call site(s)")
        return True
    else:
        print("\nNo patches applied")
        return False

def main():
    if len(sys.argv) < 4 or len(sys.argv) % 2 != 0:
        print("Usage: patch_runtime.py <elf_file> <old_func1> <new_func1> [<old_func2> <new_func2> ...]")
        print("\nExample:")
        print("  patch_runtime.py kernel.elf runtime.gcWriteBarrier2 runtime.gcWriteBarrier2")
        sys.exit(1)
    
    elf_path = sys.argv[1]
    replacements = []
    
    # Parse function pairs
    for i in range(2, len(sys.argv), 2):
        old_func = sys.argv[i]
        new_func = sys.argv[i + 1]
        replacements.append((old_func, new_func))
    
    if patch_runtime(elf_path, replacements):
        sys.exit(0)
    else:
        sys.exit(1)

if __name__ == '__main__':
    main()

File: src/mazboot/tools/patch-runtime.go
//go:build ignore
// +build ignore

// Runtime patching tool for replacing Go runtime functions.
// Scans assembly files to find .global runtime.* symbols that need patching,
// then patches ELF binaries to redirect function calls from Go runtime's weak
// symbols to our strong global implementations.
//
// Usage: go run patch-runtime.go <elf_file> <asm_dir>
//
// The tool automatically discovers which runtime symbols need patching by
// scanning .s files for .global runtime.* declarations.

package main

import (
	"debug/elf"
	"encoding/binary"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

func main() {
	if len(os.Args) < 3 {
		fmt.Fprintf(os.Stderr, "Usage: %s <elf_file> <asm_dir>\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "\nExample:\n")
		fmt.Fprintf(os.Stderr, "  %s build/mazboot/mazboot.elf src/mazboot/asm/aarch64\n", os.Args[0])
		os.Exit(1)
	}

	elfPath := os.Args[1]
	asmDir := os.Args[2]

	// Find runtime symbols that need patching by scanning .s files
	replacements := findRuntimeReplacements(asmDir)
	if len(replacements) == 0 {
		fmt.Println("No runtime symbols found to patch")
		os.Exit(0)
	}

	fmt.Printf("Found %d runtime symbol(s) to patch:\n", len(replacements))
	for old, new := range replacements {
		fmt.Printf("  %s -> %s\n", old, new)
	}

	// Patch the ELF file
	if err := patchRuntime(elfPath, replacements); err != nil {
		fmt.Fprintf(os.Stderr, "Error patching runtime: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("\nSuccessfully patched %d symbol(s)\n", len(replacements))
}

// findRuntimeReplacements scans assembly files for .global runtime.* symbols
// Returns a map of old_func -> new_func (both same name, redirecting to our implementation)
// Only includes symbols that need runtime patching (write barrier functions)
func findRuntimeReplacements(asmDir string) map[string]string {
	replacements := make(map[string]string)

	// Only patch write barrier symbols (gcWriteBarrier*)
	// Other runtime.* symbols like morestack don't need patching
	writeBarrierRe := regexp.MustCompile(`^\.global\s+(runtime\.gcWriteBarrier[234]|gcWriteBarrier)\b`)

	filepath.Walk(asmDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || !strings.HasSuffix(path, ".s") {
			return nil
		}

		content, err := ioutil.ReadFile(path)
		if err != nil {
			return nil
		}

		lines := strings.Split(string(content), "\n")
		for _, line := range lines {
			matches := writeBarrierRe.FindStringSubmatch(line)
			if len(matches) > 1 {
				symbol := matches[1]
				// For runtime.* symbols, we redirect from Go's weak version to our strong version
				// Both old and new have the same name
				replacements[symbol] = symbol
			}
		}

		return nil
	})

	return replacements
}

// patchRuntime patches an ELF binary to redirect function calls
func patchRuntime(elfPath string, replacements map[string]string) error {
	// Open ELF file
	file, err := elf.Open(elfPath)
	if err != nil {
		return fmt.Errorf("failed to open ELF file: %w", err)
	}
	defer file.Close()

	// Read the entire file
	data, err := ioutil.ReadFile(elfPath)
	if err != nil {
		return fmt.Errorf("failed to read ELF file: %w", err)
	}

	// Find .text section
	var textSection *elf.Section
	for _, section := range file.Sections {
		if section.Name == ".text" {
			textSection = section
			break
		}
	}
	if textSection == nil {
		return fmt.Errorf("could not find .text section")
	}

	// Get bin directory for target-* tools
	binDir := findBinDir()
	env := os.Environ()
	env = append(env, "PATH="+binDir+":"+os.Getenv("PATH"))

	totalPatches := 0

	// Process each replacement
	for oldFunc, newFunc := range replacements {
		// Find symbol addresses using target-nm
		oldAddr, err := findSymbolAddress(elfPath, oldFunc, env, 't') // Prefer weak/local from Go runtime
		if err != nil {
			fmt.Printf("  Warning: Could not find old symbol %s: %v\n", oldFunc, err)
			continue
		}

		newAddr, err := findSymbolAddress(elfPath, newFunc, env, 'T') // Prefer strong/global from our code
		if err != nil {
			fmt.Printf("  Warning: Could not find new symbol %s: %v\n", newFunc, err)
			continue
		}

		// Find all call sites
		callSites, err := findCallSites(elfPath, oldAddr, env)
		if err != nil {
			fmt.Printf("  Warning: Could not find call sites for %s: %v\n", oldFunc, err)
			continue
		}

		if len(callSites) == 0 {
			// Skip gcWriteBarrier if no calls found (it's optional)
			if oldFunc == "gcWriteBarrier" {
				continue
			}
			fmt.Printf("  Warning: No call sites found for %s\n", oldFunc)
			continue
		}

		// Patch each call site
		for _, callAddr := range callSites {
			if patchCallSite(data, textSection, callAddr, oldAddr, newAddr) {
				totalPatches++
			}
		}
	}

	if totalPatches > 0 {
		// Write patched file
		if err := ioutil.WriteFile(elfPath, data, 0644); err != nil {
			return fmt.Errorf("failed to write patched file: %w", err)
		}
		fmt.Printf("\nSuccessfully patched %d call site(s)\n", totalPatches)
		return nil
	}

	fmt.Println("\nNo patches applied")
	return nil
}

// findBinDir finds the bin directory containing target-* tools
func findBinDir() string {
	// Assume we're running from project root or src/mazboot
	// Try to find bin/ directory

	// Check current directory
	if _, err := os.Stat("bin"); err == nil {
		abs, _ := filepath.Abs("bin")
		return abs
	}

	// Check parent directories
	for i := 1; i <= 3; i++ {
		path := filepath.Join(strings.Repeat("../", i), "bin")
		if _, err := os.Stat(path); err == nil {
			abs, _ := filepath.Abs(path)
			return abs
		}
	}

	// Default to relative path
	return "bin"
}

// findSymbolAddress finds the address of a symbol using target-nm
func findSymbolAddress(elfPath, symbolName string, env []string, preferType byte) (uint64, error) {
	cmd := exec.Command("target-nm", elfPath)
	cmd.Env = env
	output, err := cmd.Output()
	if err != nil {
		return 0, fmt.Errorf("target-nm failed: %w", err)
	}

	var matches []struct {
		addr uint64
		typ  byte
	}

	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		parts := strings.Fields(line)
		if len(parts) >= 3 {
			addrStr := parts[0]
			symType := parts[1]
			name := strings.Join(parts[2:], " ")

			if name == symbolName && (symType[0] == 'T' || symType[0] == 't') {
				addr, err := parseHex(addrStr)
				if err != nil {
					continue
				}
				matches = append(matches, struct {
					addr uint64
					typ  byte
				}{addr, symType[0]})
			}
		}
	}

	if len(matches) == 0 {
		return 0, fmt.Errorf("symbol %s not found", symbolName)
	}

	// Prefer the specified type
	for _, m := range matches {
		if m.typ == preferType {
			return m.addr, nil
		}
	}

	// Return first match
	return matches[0].addr, nil
}

// findCallSites finds all call sites (bl instructions) that call the target address
func findCallSites(elfPath string, targetAddr uint64, env []string) ([]uint64, error) {
	cmd := exec.Command("target-objdump", "-d", elfPath)
	cmd.Env = env
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("target-objdump failed: %w", err)
	}

	var callSites []uint64
	blRe := regexp.MustCompile(`^\s*([0-9a-f]+):\s+[0-9a-f]{8}\s+bl\s+([0-9a-f]+)`)

	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		matches := blRe.FindStringSubmatch(line)
		if len(matches) >= 3 {
			callAddr, err1 := parseHex(matches[1])
			target, err2 := parseHex(matches[2])
			if err1 == nil && err2 == nil && target == targetAddr {
				callSites = append(callSites, callAddr)
			}
		}
	}

	return callSites, nil
}

// patchCallSite patches a single call site to redirect from old_target to new_target
func patchCallSite(data []byte, textSection *elf.Section, callVAddr, oldTarget, newTarget uint64) bool {
	// Calculate file offset
	fileOffset := int64(textSection.Offset) + int64(callVAddr-textSection.Addr)

	if fileOffset < 0 || fileOffset >= int64(len(data)) {
		fmt.Printf("  Warning: Invalid file offset 0x%x for call at 0x%x\n", fileOffset, callVAddr)
		return false
	}

	// Read current instruction
	currentInsn := binary.LittleEndian.Uint32(data[fileOffset:])

	// Verify it's a bl instruction (opcode 0x94000000)
	if (currentInsn & 0xfc000000) != 0x94000000 {
		fmt.Printf("  Warning: Instruction at 0x%x is not a bl instruction: 0x%x\n", callVAddr, currentInsn)
		return false
	}

	// Decode current target to verify
	imm26 := currentInsn & 0x3ffffff
	// Sign extend 26-bit immediate
	if imm26&0x2000000 != 0 {
		imm26 |= 0xfc000000 // Sign extend
	}
	currentOffset := int64(imm26) << 2
	currentTarget := callVAddr + uint64(currentOffset)

	if currentTarget != oldTarget {
		// Allow some address differences (LOAD_ADDRESS, RAM_BASE offsets)
		diff := currentTarget
		if currentTarget > oldTarget {
			diff = currentTarget - oldTarget
		} else {
			diff = oldTarget - currentTarget
		}
		if diff != 0x200000 && diff != 0x40000000 && diff != 0x400000000 {
			fmt.Printf("  Warning: Call at 0x%x targets 0x%x, expected 0x%x\n", callVAddr, currentTarget, oldTarget)
		}
	}

	// Encode new instruction
	newOffset := int64(newTarget) - int64(callVAddr)
	if newOffset%4 != 0 {
		fmt.Printf("  Error: Target address must be 4-byte aligned (offset=%d)\n", newOffset)
		return false
	}
	if newOffset < -0x8000000 || newOffset > 0x7ffffff {
		fmt.Printf("  Error: Branch offset out of range: %d (max ±128MB)\n", newOffset)
		return false
	}

	imm26 = uint32((newOffset >> 2) & 0x3ffffff)
	newInsn := 0x94000000 | imm26

	// Write the new instruction
	binary.LittleEndian.PutUint32(data[fileOffset:], newInsn)
	return true
}

// parseHex parses a hexadecimal string
func parseHex(s string) (uint64, error) {
	var val uint64
	_, err := fmt.Sscanf(s, "%x", &val)
	return val, err
}



